UnityProject/Assets/Scripts/Weapons/Gun.cs

```diff
@@ -0,0 +1,889 @@
+				//populate with a full internal mag on spawn
+			}
+
+			if (ammoPrefab == null)
+			{
+				Debug.LogError($"{gameObject.name} magazine prefab was null, cannot auto-populate.");
+				return;
+			}
+
+			//populate with a full external mag on spawn
+			Logger.LogTraceFormat("Auto-populate external magazine for {0}", Category.Inventory, name);
+			Inventory.ServerAdd(Spawn.ServerPrefab(ammoPrefab).GameObject, magSlot);
+		}
+
+		public void OnInventoryMoveServer(InventoryMove info)
+		{
+			if (info.ToPlayer != null)
+			{
+				serverHolder = info.ToPlayer.gameObject;
+				shooterRegisterTile = serverHolder.GetComponent<RegisterTile>();
+			}
+			else
+			{
+				serverHolder = null;
+				shooterRegisterTile = null;
+			}
+		}
+
+		#endregion
+
+		#region Interaction
+
+		public bool WillInteract(AimApply interaction, NetworkSide side)
+		{
+			if (!DefaultWillInteract.Default(interaction, side)) return false;
+			if (CurrentMagazine == null)
+			{
+				PlayEmptySFX();
+				if (interaction.Performer != PlayerManager.LocalPlayer)
+				{
+					Logger.LogTrace("Server rejected shot - No magazine being loaded", Category.Firearms);
+				}
+				return false;
+			}
+
+			//note: fire count down is only checked local player side, as server processes all the shots in a queue
+			//anyway so client cannot exceed that firing rate no matter what. If we validate firing rate server
+			//side at the moment of interaction, it will reject client's shots because of lag between server / client
+			//firing countdown
+			if (Projectile != null && CurrentMagazine.ClientAmmoRemains <= 0 && (interaction.Performer != PlayerManager.LocalPlayer || FireCountDown <= 0))
+			{
+				if (SmartGun && allowMagazineRemoval) // smartGun is forced off when using an internal magazine
+				{
+					RequestUnload(CurrentMagazine);
+					OutOfAmmoSFX();
+				}
+				else
+				{
+					PlayEmptySFX();
+				}
+				if (interaction.Performer != PlayerManager.LocalPlayer)
+				{
+					Logger.LogTrace("Server rejected shot - out of ammo", Category.Firearms);
+				}
+				return false;
+			}
+
+			if (Projectile != null && CurrentMagazine.ClientAmmoRemains > 0 && (interaction.Performer != PlayerManager.LocalPlayer || FireCountDown <= 0))
+			{
+				if (WeaponType == WeaponType.Burst)
+				{
+					//being held and is a burst weapon, check how many shots have been fired our the current burst
+					if (currentBurstCount < burstCount)
+					{
+						//we have shot less then the max allowed shots in our current burst, increase and fire again
+						currentBurstCount++;
+						return true;
+					}
+					else if (currentBurstCount >= burstCount)
+					{
+						//we have shot the max allowed shots in our current burst, start cooldown and then fire the first shot
+						WaitFor.Seconds((float)burstCooldown);
+						currentBurstCount = 1;
+						return true;
+					}
+				}
+				else if (interaction.MouseButtonState == MouseButtonState.PRESS)
+				{
+					if (currentBurstCount != 0)
```

Comment:
Codacy found an issue: [Do not check floating point inequality with exact values, use a range instead.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6280689)